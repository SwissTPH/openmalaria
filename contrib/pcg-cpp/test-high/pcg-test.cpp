/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
