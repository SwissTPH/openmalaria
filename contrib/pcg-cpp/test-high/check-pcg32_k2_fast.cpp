#define RNG pcg32_k2_fast
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

