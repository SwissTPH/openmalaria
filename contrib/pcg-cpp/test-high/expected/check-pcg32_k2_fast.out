pcg32_k2_fast:
      -  result:      32-bit unsigned int
      -  period:      2^128
      -  size:        16 bytes

Round 1:
  32bit: 0xe30b7cba 0x6af364b0 0x27de7b5d 0xcc1c87c3 0x5cd80226 0xfec28ee6
  Again: 0xe30b7cba 0x6af364b0 0x27de7b5d 0xcc1c87c3 0x5cd80226 0xfec28ee6
  Coins: HTTHTHTHHTHHTHTTTHTHTHTHTHTHHHTHTHTTTTTTHTTTTTTHTTTHHHHTHHTTTHHTT
  Rolls: 1 3 4 5 3 6 2 4 2 5 1 6 4 5 5 4 2 4 3 4 5 1 2 1 4 6 5 4 2 2 5 1 4
   -->   rolling dice used 33 random numbers
  Cards: Jc 7c 4c Kh As 6s Qc Qh Qd 3d Th 8d Qs 5s 5d Kd 7h 8h Ah 4d Jh 5h
	 Kc 2h Tc 7d 9h 6c 4s 8c 8s 9s 9d 2d 6h 5c Jd 3h 4h 2s 3s Ks Td Ts
	 3c 9c 7s 2c 6d Js Ac Ad

Round 2:
  32bit: 0x162d493a 0x03abb49b 0x00ef59f5 0x7da5ed07 0xb9008c9c 0x1cf59ea5
  Again: 0x162d493a 0x03abb49b 0x00ef59f5 0x7da5ed07 0xb9008c9c 0x1cf59ea5
  Coins: THTTTTTTHTHHHHTHTTHHTTHHHHTTHHHHHTHTHHTTTHTHHHHTHTTTTTTTTTTTHTHTT
  Rolls: 6 1 3 1 1 3 5 3 2 2 3 5 3 1 4 5 1 6 4 2 4 6 2 4 6 3 2 3 4 1 6 2 5
   -->   rolling dice used 33 random numbers
  Cards: 8h 3h 7s 9c 9d Ac Jh 7d 6h Ks 3s 7c 2s Ad Th 5c 4h Kh Tc Kd Jc 6s
	 Qd 2d 3d Qc Qs Qh 5s 6d 5h As 9h 2h Ts 6c 4d Kc Js 3c 9s Jd 7h 4c
	 Td 8d 2c 5d 8s Ah 8c 4s

Round 3:
  32bit: 0xc3981ded 0x898126c2 0x6bf9f962 0x0644d3dc 0x1e45ccb6 0x9d2ecaf2
  Again: 0xc3981ded 0x898126c2 0x6bf9f962 0x0644d3dc 0x1e45ccb6 0x9d2ecaf2
  Coins: THTHHHHHTTHTHHHHHTHTHHHHTTTTHHHHHHTTTTHHHTTHHHTHTTHTTHHHHTHTHHTHH
  Rolls: 4 2 5 5 6 6 6 2 1 6 2 2 4 3 4 3 1 5 2 2 6 6 1 4 6 5 1 3 2 2 1 3 6
   -->   rolling dice used 33 random numbers
  Cards: 3c 8s 7d 5s Qc 4h Kc 5h 4s Kd 3d 3h Ah Qs 4d Jd 8d Ac 7c 2c 6s 7s
	 Qh Ks 6d 9d 2d 6c Td Jh 4c Js Tc 3s Ad Qd 5c Th Jc Kh 9c 2s 7h As
	 9h 8c 6h 2h 5d Ts 9s 8h

Round 4:
  32bit: 0x8575f7de 0x96024950 0x9d505d02 0x5623edf7 0xa615206d 0xe9da85d2
  Again: 0x8575f7de 0x96024950 0x9d505d02 0x5623edf7 0xa615206d 0xe9da85d2
  Coins: HHHHHTTTHHHTTHHHTTTTHHHTTTTTHHTTHHTHHTHHHHTTTTTTTTHHTTHHHHTHHTHTH
  Rolls: 6 1 2 6 2 5 1 4 1 5 4 5 5 5 1 3 1 5 6 6 2 2 1 2 5 5 5 6 4 2 1 5 1
   -->   rolling dice used 33 random numbers
  Cards: 7d 3h Qs 9h 5h Jh 9s Js Kd Jd Kh 2d 2s 8h Qd Qh Ac 9c Kc Jc 8d As
	 4c 3d 5s 5d 5c 8s 7s Qc 2c Td Ks 4s 7h Ad 2h 6h 6d 3c 6s 6c 9d Tc
	 Ah 7c Ts 4d 3s Th 8c 4h

Round 5:
  32bit: 0x3c64d5ba 0x6de15115 0xd50081e4 0x4e0d6399 0x34b5cfc7 0x8f7825ab
  Again: 0x3c64d5ba 0x6de15115 0xd50081e4 0x4e0d6399 0x34b5cfc7 0x8f7825ab
  Coins: TTTTTTTHHHTTTHTTTTTTTTHHHTHTTHTTTHHTTTHTTTTHHTTHTTHTTHHTTTTHHTHTH
  Rolls: 4 2 5 5 2 1 2 4 5 4 5 1 3 6 3 3 3 1 5 3 2 3 1 6 5 5 1 5 6 6 1 3 2
   -->   rolling dice used 33 random numbers
  Cards: 5h Kd Qd Qs Jc 3h 6s 7d 4c 4d Ts Tc 3d 7s 6h 3c 2d 9d 6d 6c 9s 2s
	 7h 8c 5c Ks 5d 7c 4s 8d Qh Td Ah Js As Kc 9h 9c Th 3s 2h Qc 4h Ac
	 5s Kh 8h 8s 2c Jd Ad Jh

